### PR TITLE
HashJoin: use unordered_map instead of cuckoo hashing

### PR DESCRIPTION
**Pull request by Bouncner**
This PR changes the hash table used in the hash join to a simple `std::unordered_map`.

The main reason was that the concurrent WIP PR #1049 triggered the sometimes occurring problems with our cuckoo table much more frequently.

On top of that and for me also surprisingly, the `std::unordered_map` is even faster:
```
Master:
./cmake_release/hyriseMicroBenchmarks --benchmark_filter=JoinHash\* --benchmark_repetitions=10
-----------------------------------------------------------------------------
Benchmark                                Time                 CPU Iterations
-----------------------------------------------------------------------------
BM_Join_Small<JoinHash>_mean       1,697,778 ns    1,689,193 ns        387
BM_Join_Skewed<JoinHash>_mean    773,188,627 ns  768,600,600 ns          1
BM_Join_Big<JoinHash>_mean        93,289,491 ns   92,909,533 ns          6

This PR:
-----------------------------------------------------------------------------
Benchmark                                Time                 CPU Iterations
-----------------------------------------------------------------------------
BM_Join_Small<JoinHash>_mean       1,494,822 ns    1,480,951 ns        464
BM_Join_Skewed<JoinHash>_mean    613,261,750 ns  609,473,200 ns          1
BM_Join_Big<JoinHash>_mean        85,425,641 ns   85,003,443 ns          7
```

This PR removes a hand-written hash table with a std-based hash table, which allows us to test other hash map implementations (sharing the same interface) that are (i) faster than std::unordered_map and (ii) more space-efficient.

### CODE REVIEW

**Comment by mrks** [comment]
Could you please also provide numbers for TPC-H?

**Comment by Bouncner** [comment]
Currently running ...

**Comment by Bouncner** [comment]
And results for `scale=0.1`:
```
+-----------+---------------------+------+--------------------+------+--------+-----------------------------------------------+
| Benchmark | prev. iter/s        | runs | new iter/s         | runs | change |               p-value (significant if <0.001) |
+-----------+---------------------+------+--------------------+------+--------+-----------------------------------------------+
| TPC-H 1   | 2.4382209777832     | 13   | 2.46756267547607   | 13   | +1%    |                   (run time too short) 0.6414 |
| TPC-H 2   | 13.8464708328247    | 70   | 13.652416229248    | 69   | -1%    |                   (run time too short) 0.3529 |
| TPC-H 3   | 13.7707166671753    | 69   | 13.8171052932739   | 70   | +0%    |                   (run time too short) 0.8336 |
| TPC-H 4   | 0.327766120433807   | 2    | 0.412352234125137  | 3    | +26%   | (run time too short) (not enough runs) 0.0034 |
| TPC-H 5   | 9.81660842895508    | 50   | 9.79790782928467   | 49   | -0%    |                   (run time too short) 0.9079 |
| TPC-H 6   | 40.3500022888184    | 202  | 39.9615631103516   | 200  | -1%    |                   (run time too short) 0.2987 |
| TPC-H 7   | 2.90275120735168    | 15   | 2.97938919067383   | 15   | +3%    |                   (run time too short) 0.1686 |
| TPC-H 8   | 2.4810950756073     | 13   | 2.31846404075623   | 12   | -7%    |                   (run time too short) 0.0036 |
| TPC-H 9   | 1.43333959579468    | 8    | 1.34580373764038   | 7    | -6%    | (run time too short) (not enough runs) 0.0122 |
| TPC-H 10  | 9.56744575500488    | 48   | 9.83222484588623   | 50   | +3%    |                   (run time too short) 0.0928 |
| TPC-H 11  | 29.9462070465088    | 150  | 28.6363754272461   | 144  | -4%    |                   (run time too short) 0.0000 |
| TPC-H 12  | 13.2450475692749    | 67   | 13.6787977218628   | 69   | +3%    |                   (run time too short) 0.0066 |
| TPC-H 13  | 12.8717784881592    | 65   | 11.2632703781128   | 57   | -12%   |                   (run time too short) 0.0000 |
| TPC-H 14  | 28.742431640625     | 144  | 29.2504615783691   | 147  | +2%    |                   (run time too short) 0.0556 |
| TPC-H 15  | 0.215460941195488   | 2    | 0.223195970058441  | 2    | +4%    | (run time too short) (not enough runs) 0.3505 |
| TPC-H 16  | 6.56419467926025    | 33   | 6.17109727859497   | 31   | -6%    |                   (run time too short) 0.0100 |
| TPC-H 17  | 1.32866430282593    | 7    | 1.30945432186127   | 7    | -1%    | (run time too short) (not enough runs) 0.2780 |
| TPC-H 18  | 0.0222702454775572  | 1    | 0.023688118904829  | 1    | +6%    |    (run time too short) (not enough runs) nan |
| TPC-H 19  | 1.33674013614655    | 7    | 1.2693567276001    | 7    | -5%    | (run time too short) (not enough runs) 0.3823 |
| TPC-H 20  | 0.00526655884459615 | 1    | 0.0053626848384738 | 1    | +2%    |                         (not enough runs) nan |
| TPC-H 21  | 0.230141177773476   | 2    | 0.234685733914375  | 2    | +2%    | (run time too short) (not enough runs) 0.6825 |
| TPC-H 22  | 0.290742963552475   | 2    | 0.295315533876419  | 2    | +2%    | (run time too short) (not enough runs) 0.4553 |
| average   |                     |      |                    |      | +0%    |                                               |
+-----------+---------------------+------+--------------------+------+--------+-----------------------------------------------+
```

File: src/lib/operators/join_hash.cpp [hunk 1]
```diff
@@ -119,12 +121,13 @@ struct RadixContainer {
...
 template <typename LeftType, typename HashedType>
-std::vector<std::optional<HashTable<HashedType>>> build(const RadixContainer<LeftType>& radix_container) {
+std::vector<std::optional<std::unordered_map<HashedType, boost::variant<RowID, PosList>>>> build(
+    const RadixContainer<LeftType>& radix_container) {
   /*
   NUMA notes:
   The hashtables for each partition P should also reside on the same node as the two vectors leftP and rightP.
   */
-  std::vector<std::optional<HashTable<HashedType>>> hashtables;
+  std::vector<std::optional<std::unordered_map<HashedType, boost::variant<RowID, PosList>>>> hashtables;
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: `using` could save us this repetition.

**Comment by mrks** [comment]
Other than the missing `using`, I don't have anything to comment on the code side. As the TPC-H benchmarks are way less convincing than the previous benchmarks, I wonder if we should discuss this tomorrow.

**Comment by Bouncner** [comment]
I will run some longer running TPC-H benchmarks to get a more accurate picture. Is nemea the best place for that?

File: src/lib/operators/join_hash.cpp (hunk 1 above)
Comment (Bouncner) [reply src/lib/operators/join_hash.cpp]: Not sure yet how to combine a template type with using. :-/
Let's discuss tomorrow.

**Comment by Bouncner** [comment]
Here a longer running TPC-H benchmark. Scale factor 0.1.

```
+-----------+---------------------+------+---------------------+------+--------+---------------------------------+
| Benchmark | prev. iter/s        | runs | new iter/s          | runs | change | p-value (significant if <0.001) |
+-----------+---------------------+------+---------------------+------+--------+---------------------------------+
| TPC-H 1   | 2.73134922981262    | 1000 | 2.70583200454712    | 1000 | -1%    |                          0.0000 |
| TPC-H 2   | 20.9591464996338    | 1000 | 22.0250205993652    | 1000 | +5%    |     (run time too short) 0.0000 |
| TPC-H 3   | 16.2167778015137    | 1000 | 22.2300300598145    | 1000 | +37%   |     (run time too short) 0.0000 |
| TPC-H 4   | 0.31084680557251    | 224  | 0.310456722974777   | 224  | -0%    |                          0.5524 |
| TPC-H 5   | 12.1663751602173    | 1000 | 15.6403093338013    | 1000 | +29%   |                          0.0000 |
| TPC-H 6   | 49.6075668334961    | 1000 | 49.271183013916     | 1000 | -1%    |     (run time too short) 0.0074 |
| TPC-H 7   | 4.29073572158813    | 1000 | 4.75418281555176    | 1000 | +11%   |                          0.0000 |
| TPC-H 8   | 3.10401010513306    | 1000 | 3.23372077941895    | 1000 | +4%    |                          0.0000 |
| TPC-H 9   | 1.78640758991241    | 1000 | 1.92059361934662    | 1000 | +8%    |                          0.0000 |
| TPC-H 10  | 14.4318647384644    | 1000 | 15.4717359542847    | 1000 | +7%    |                          0.0000 |
| TPC-H 11  | 40.8472366333008    | 1000 | 44.0052032470703    | 1000 | +8%    |     (run time too short) 0.0000 |
| TPC-H 12  | 17.6362342834473    | 1000 | 19.3214855194092    | 1000 | +10%   |     (run time too short) 0.0000 |
| TPC-H 13  | 13.5514631271362    | 1000 | 17.6978759765625    | 1000 | +31%   |     (run time too short) 0.0000 |
| TPC-H 14  | 39.5775451660156    | 1000 | 40.5096244812012    | 1000 | +2%    |     (run time too short) 0.0000 |
| TPC-H 15  | 0.268731832504272   | 194  | 0.27754470705986    | 200  | +3%    |                          0.0000 |
| TPC-H 16  | 6.81569290161133    | 1000 | 6.89850330352783    | 1000 | +1%    |                          0.0000 |
| TPC-H 17  | 1.3637033700943     | 982  | 1.36338913440704    | 982  | -0%    |                          0.9027 |
| TPC-H 18  | 0.0334157757461071  | 25   | 0.0330488756299019  | 24   | -1%    |                          0.0000 |
| TPC-H 19  | 1.73480129241943    | 1000 | 1.76322031021118    | 1000 | +2%    |                          0.0000 |
| TPC-H 20  | 0.00561590725556016 | 5    | 0.00562030030414462 | 5    | +0%    |        (not enough runs) 0.6793 |
| TPC-H 21  | 0.184361517429352   | 133  | 0.185666546225548   | 134  | +1%    |                          0.0002 |
| TPC-H 22  | 0.306500941514969   | 221  | 0.306894689798355   | 221  | +0%    |                          0.5168 |
| average   |                     |      |                     |      | +7%    |                                 |
+-----------+---------------------+------+---------------------+------+--------+---------------------------------+
```

Max loses of -1% (pretty sure it's not that PR) and improvements of up to 37%.

**Comment by mrks** [comment]
What's up with the `murmur_hash.*` files? Do we still need those?

**Comment by Bouncner** [comment]
We use murmur hashing for the radix cluster calculation (does not really hurt to use std::hash here for our simple values). So it's used but could be removed.

Only thing that might use murmur in the future are the filters from @ArneMayer . Can you tell us something about that, @ArneMayer ?

**Comment by mrks** [comment]
No need to remove it, just wanted to check

**Comment by Bouncner** [comment]
Whooopsi ... it's new branch.
I just pushed the changes with `using ...`. Shall I open a new PR for that? It's just 4 lines ...

**Comment by ArneMayer** [comment]
@Bouncner The filters bring their own hashing function.

**Pull request merged**